src/envoy/http/authnv2/filter.cc

```diff
@@ -0,0 +1,236 @@
+    // support string_list (https://github.com/istio/istio/issues/2802) For
+    // now, just use the first value.
+    const auto& aud_values = (*claims)[kJwtAudienceKey].list_value().values();
+    if (!aud_values.empty()) {
+      setKeyValue(authn_data,
+                  istio::utils::AttributeName::kRequestAuthAudiences,
+                  aud_values[0].string_value());
+    }
+  }
+
+  // request.auth.principal
+  if (claims->find("iss") != claims->end() &&
+      claims->find("sub") != claims->end()) {
+    setKeyValue(
+        authn_data, istio::utils::AttributeName::kRequestAuthPrincipal,
+        (*claims)["iss"].list_value().values().Get(0).string_value() + "/" +
+            (*claims)["sub"].list_value().values().Get(0).string_value());
+  }
+
+  // request.auth.audiences
+  if (claims->find("azp") != claims->end()) {
+    setKeyValue(authn_data, istio::utils::AttributeName::kRequestAuthPresenter,
+                (*claims)["azp"].list_value().values().Get(0).string_value());
+  }
+
+  // request.auth.claims
+  (*(*authn_data
+          .mutable_fields())[istio::utils::AttributeName::kRequestAuthClaims]
+        .mutable_struct_value())
+      .MergeFrom(claim_structs);
+
+  // request.auth.raw_claims
+  setKeyValue(authn_data, istio::utils::AttributeName::kRequestAuthRawClaims,
+              jwt);
+
+  return true;
+}
+
+// Returns true if the peer identity can be extracted from the underying
+// certificate used for mTLS.
+bool ProcessMtls(const Network::Connection* connection,
+                 ProtobufWkt::Struct& authn_data) {
+  std::string peer_principle;
+  if (connection->ssl() == nullptr ||
+      !connection->ssl()->peerCertificatePresented()) {
+    return false;
+  }
+  Utils::GetPrincipal(connection, true, &peer_principle);
+  if (peer_principle != "") {
+    setKeyValue(authn_data, istio::utils::AttributeName::kSourcePrincipal,
+                peer_principle);
+  }
+  return true;
+}
+
+std::string AuthnV2Filter::extractJwtFromMetadata(
+    const envoy::api::v2::core::Metadata& metadata, std::string* jwt_payload) {
+  std::string issuer_selected = "";
+  auto filter_it = metadata.filter_metadata().find(
+      Extensions::HttpFilters::HttpFilterNames::get().JwtAuthn);
+  if (filter_it == metadata.filter_metadata().end()) {
+    return "";
+  }
+  const ::google::protobuf::Struct& jwt_metadata = filter_it->second;
+  // Iterate over the envoy.jwt metadata, which is indexed by the issuer.
+  // For multiple JWT, we only select on of them, the first one lexically
+  // sorted.
+  for (const auto& entry : jwt_metadata.fields()) {
+    const std::string& issuer = entry.first;
+    if (issuer_selected == "" || issuer_selected.compare(issuer) > 0) {
+      issuer_selected = issuer;
+    }
+  }
+  if (issuer_selected != "") {
+    const auto& jwt_entry = jwt_metadata.fields().find(issuer_selected);
+    Protobuf::util::MessageToJsonString(jwt_entry->second.struct_value(),
```

Comment:
You convert the struct to string here and later convert the string back to struct in `ProcessJwt`. I think we could use the struct directly in the 2 functions.

Reply:
This is how it's implemented now. 

1. [`getJwtFromEnvoyFilter`](https://github.com/istio/proxy/blob/master/src/envoy/http/authn/filter_context.cc#L84) calling proto to string serialization
1. passing as proto to validate, in [`validateJwt`[https://github.com/istio/proxy/blob/master/src/envoy/http/authn/authenticator_base.cc#L124)

Plus the extract string does have some string split of the keys, https://github.com/istio/proxy/blob/master/src/envoy/http/authn/authn_utils.cc#L48

Reply:
take another closer look, and can't understand the reason why old impl convert it back and forth...
i'll take a look see if we can use the struct directly without converting back and forth.

Reply:
It's a tech debt that should be cleaned from the old impl as well.